File: src/transfers/TransferMarket.cs
using ElevenLegends.Data.Enums;
using ElevenLegends.Data.Models;

namespace ElevenLegends.Transfers;

/// <summary>
/// Handles transfer operations: buy, sell, loan in, loan out.
/// Validates squad size rules (min 14, max 22, at least 1 GK).
/// </summary>
public static class TransferMarket
{
    public const int MinSquadSize = 14;
    public const int MaxSquadSize = 22;

    /// <summary>
    /// Returns players from other clubs that are available for purchase.
    /// </summary>
    public static List<(Player Player, Club Club, decimal Price)> GetAvailablePlayers(
        IReadOnlyList<Club> clubs, int excludeClubId)
    {
        var available = new List<(Player, Club, decimal)>();

        foreach (var club in clubs)
        {
            if (club.Id == excludeClubId) continue;
            if (club.Team.Players.Count <= MinSquadSize) continue;

            var starterSet = new HashSet<int>(club.Team.StartingLineup);
            foreach (var player in club.Team.Players)
            {
                if (starterSet.Contains(player.Id)) continue;
                decimal price = PlayerValuation.Calculate(player, club.Reputation);
                available.Add((player, club, price));
            }
        }

        return available;
    }

    /// <summary>
    /// Returns players from a club that can be sold.
    /// </summary>
    public static List<(Player Player, decimal Price)> GetSellablePlayers(Club club)
    {
        if (club.Team.Players.Count <= MinSquadSize)
            return [];

        var results = new List<(Player, decimal)>();
        foreach (var player in club.Team.Players)
        {
            if (!CanRemovePlayer(club, player)) continue;
            decimal price = PlayerValuation.Calculate(player, club.Reputation);
            results.Add((player, price));
        }
        return results;
    }

    /// <summary>
    /// Returns players from other clubs available for loan.
    /// </summary>
    public static List<(Player Player, Club Club, decimal Fee)> GetLoanablePlayersIn(
        IReadOnlyList<Club> clubs, int excludeClubId)
    {
        var available = new List<(Player, Club, decimal)>();

        foreach (var club in clubs)
        {
            if (club.Id == excludeClubId) continue;
            if (club.Team.Players.Count <= MinSquadSize) continue;

            var starterSet = new HashSet<int>(club.Team.StartingLineup);
            foreach (var player in club.Team.Players)
            {
                if (starterSet.Contains(player.Id)) continue;
                decimal fee = Math.Round(PlayerValuation.Calculate(player, club.Reputation) * 0.3m / 1000m) * 1000m;
                available.Add((player, club, Math.Max(2_000m, fee)));
            }
        }

        return available;
    }

    /// <summary>
    /// Executes a player purchase. Returns true if successful.
    /// </summary>
    public static bool ExecuteBuy(Club buyer, Club seller, Player player, decimal price)
    {
        if (buyer.Team.Players.Count >= MaxSquadSize) return false;
        if (buyer.Balance < price) return false;
        if (!seller.Team.Players.Contains(player)) return false;
        if (!CanRemovePlayer(seller, player)) return false;

        buyer.Balance -= price;
        seller.Balance += price;

        RemovePlayerFromClub(seller, player);
        AddPlayerToClub(buyer, player);

        return true;
    }

    /// <summary>
    /// Executes a player sale. Returns true if successful.
    /// </summary>
    public static bool ExecuteSell(Club seller, Club buyer, Player player, decimal price)
    {
        return ExecuteBuy(buyer, seller, player, price);
    }

    /// <summary>
    /// Executes a loan-in: player moves temporarily to the host club.
    /// Returns the LoanRecord if successful.
    /// </summary>
    public static LoanRecord? ExecuteLoanIn(Club host, Club source, Player player, decimal fee)
    {
        if (host.Team.Players.Count >= MaxSquadSize) return null;
        if (host.Balance < fee) return null;
        if (!source.Team.Players.Contains(player)) return null;
        if (!CanRemovePlayer(source, player)) return null;

        host.Balance -= fee;
        source.Balance += fee;

        RemovePlayerFromClub(source, player);
        AddPlayerToClub(host, player);

        return new LoanRecord
        {
            PlayerId = player.Id,
            PlayerName = player.Name,
            OriginClubId = source.Id,
            HostClubId = host.Id
        };
    }

    /// <summary>
    /// Executes a loan-out: player is sent to another club.
    /// Returns the LoanRecord if successful.
    /// </summary>
    public static LoanRecord? ExecuteLoanOut(Club source, Club host, Player player)
    {
        if (host.Team.Players.Count >= MaxSquadSize) return null;
        if (!source.Team.Players.Contains(player)) return null;
        if (!CanRemovePlayer(source, player)) return null;

        RemovePlayerFromClub(source, player);
        AddPlayerToClub(host, player);

        return new LoanRecord
        {
            PlayerId = player.Id,
            PlayerName = player.Name,
            OriginClubId = source.Id,
            HostClubId = host.Id
        };
    }

    /// <summary>
    /// Adds a free agent / youth recruit to a club.
    /// </summary>
    public static bool AddFreeAgent(Club club, Player player, decimal fee = 0m)
    {
        if (club.Team.Players.Count >= MaxSquadSize) return false;
        if (fee > 0 && club.Balance < fee) return false;

        if (fee > 0)
            club.Balance -= fee;

        AddPlayerToClub(club, player);
        return true;
    }

    /// <summary>
    /// Checks if a player can be removed from a club without violating rules.
    /// </summary>
    public static bool CanRemovePlayer(Club club, Player player)
    {
        if (club.Team.Players.Count <= MinSquadSize) return false;

        if (player.PrimaryPosition == Position.GK)
        {
            int gkCount = club.Team.Players.Count(p => p.PrimaryPosition == Position.GK);
            if (gkCount <= 1) return false;
        }

        return true;
    }

    private static void RemovePlayerFromClub(Club club, Player player)
    {
        var newPlayers = club.Team.Players.Where(p => p.Id != player.Id).ToList();
        var newLineup = club.Team.StartingLineup.Where(id => id != player.Id).ToList();

        // If we removed a starter, fill from reserves
        if (club.Team.StartingLineup.Contains(player.Id) && newPlayers.Count >= 11)
        {
            var lineupSet = new HashSet<int>(newLineup);
            var reserve = newPlayers.FirstOrDefault(p => !lineupSet.Contains(p.Id));
            if (reserve != null)
                newLineup.Add(reserve.Id);
        }

        club.Team = club.Team with { Players = newPlayers, StartingLineup = newLineup };
    }

    private static void AddPlayerToClub(Club club, Player player)
    {
        var newPlayers = club.Team.Players.ToList();
        newPlayers.Add(player);

        var newLineup = club.Team.StartingLineup.ToList();

        if (newLineup.Count < 11)
            newLineup.Add(player.Id);

        club.Team = club.Team with { Players = newPlayers, StartingLineup = newLineup };
    }
}

File: engine/transfers/transfer-market.ts
import { Position } from "@/engine/enums";
import type { Club, LoanRecord, Player, Team } from "@/engine/models";
import { calculateWeeklySalary } from "@/engine/economy/economy-processor";
import { calculate as calculateValuation } from "./player-valuation";

/**
 * Handles transfer operations: buy, sell, loan in, loan out.
 * Validates squad size rules (min 14, max 22, at least 1 GK).
 */

export const MIN_SQUAD_SIZE = 14;
export const MAX_SQUAD_SIZE = 22;

/**
 * Returns ALL players from other clubs with transfer prices.
 * Starters have a release clause multiplier (3-5x).
 * Clubs at minimum squad can only sell non-essential players.
 */
export function getAvailablePlayers(
  clubs: readonly Club[],
  excludeClubId: number,
): Array<{ player: Player; club: Club; price: number }> {
  const available: Array<{ player: Player; club: Club; price: number }> = [];

  for (const club of clubs) {
    if (club.id === excludeClubId) continue;

    const starterSet = new Set(club.team.startingLineup);
    for (const player of club.team.players) {
      const basePrice = calculateValuation(player, club.reputation);

      // Starters have a massive release clause
      const isStarter = starterSet.has(player.id);
      const price = isStarter ? basePrice * 4 : basePrice;

      // Can't buy if it would leave club below minimum
      if (!canRemovePlayer(club, player)) continue;

      available.push({ player, club, price });
    }
  }

  return available;
}

/**
 * Returns players from a club that can be sold.
 */
export function getSellablePlayers(
  club: Club,
): Array<{ player: Player; price: number }> {
  if (club.team.players.length <= MIN_SQUAD_SIZE) return [];

  const results: Array<{ player: Player; price: number }> = [];
  for (const player of club.team.players) {
    if (!canRemovePlayer(club, player)) continue;
    const price = calculateValuation(player, club.reputation);
    results.push({ player, price });
  }
  return results;
}

/**
 * Returns players from other clubs available for loan.
 */
export function getLoanablePlayersIn(
  clubs: readonly Club[],
  excludeClubId: number,
): Array<{ player: Player; club: Club; fee: number }> {
  const available: Array<{ player: Player; club: Club; fee: number }> = [];

  for (const club of clubs) {
    if (club.id === excludeClubId) continue;
    if (club.team.players.length <= MIN_SQUAD_SIZE) continue;

    const starterSet = new Set(club.team.startingLineup);
    for (const player of club.team.players) {
      if (starterSet.has(player.id)) continue;
      const rawFee =
        Math.round(
          (calculateValuation(player, club.reputation) * 0.3) / 1000,
        ) * 1000;
      const fee = Math.max(2_000, rawFee);
      available.push({ player, club, fee });
    }
  }

  return available;
}

/**
 * Minimum balance to keep after a purchase (reserve for 1 week of salary).
 */
export function salaryReserve(club: Club): number {
  return calculateWeeklySalary(club);
}

/**
 * Executes a player purchase. Returns true if successful.
 */
export function executeBuy(
  buyer: Club,
  seller: Club,
  player: Player,
  price: number,
): boolean {
  if (buyer.team.players.length >= MAX_SQUAD_SIZE) return false;
  const reserve = salaryReserve(buyer);
  if (buyer.balance - price < reserve) return false;
  if (!seller.team.players.some((p) => p.id === player.id)) return false;
  if (!canRemovePlayer(seller, player)) return false;

  buyer.balance -= price;
  seller.balance += price;

  removePlayerFromClub(seller, player);
  addPlayerToClub(buyer, player);

  return true;
}

/**
 * Executes a player sale. Returns true if successful.
 */
export function executeSell(
  seller: Club,
  buyer: Club,
  player: Player,
  price: number,
): boolean {
  return executeBuy(buyer, seller, player, price);
}

/**
 * Executes a loan-in: player moves temporarily to the host club.
 * Returns the LoanRecord if successful.
 */
export function executeLoanIn(
  host: Club,
  source: Club,
  player: Player,
  fee: number,
): LoanRecord | null {
  if (host.team.players.length >= MAX_SQUAD_SIZE) return null;
  if (host.balance < fee) return null;
  if (!source.team.players.some((p) => p.id === player.id)) return null;
  if (!canRemovePlayer(source, player)) return null;

  host.balance -= fee;
  source.balance += fee;

  removePlayerFromClub(source, player);
  addPlayerToClub(host, player);

  return {
    playerId: player.id,
    playerName: player.name,
    originClubId: source.id,
    hostClubId: host.id,
  };
}

/**
 * Executes a loan-out: player is sent to another club.
 * Returns the LoanRecord if successful.
 */
export function executeLoanOut(
  source: Club,
  host: Club,
  player: Player,
): LoanRecord | null {
  if (host.team.players.length >= MAX_SQUAD_SIZE) return null;
  if (!source.team.players.some((p) => p.id === player.id)) return null;
  if (!canRemovePlayer(source, player)) return null;

  removePlayerFromClub(source, player);
  addPlayerToClub(host, player);

  return {
    playerId: player.id,
    playerName: player.name,
    originClubId: source.id,
    hostClubId: host.id,
  };
}

/**
 * Adds a free agent / youth recruit to a club.
 */
export function addFreeAgent(
  club: Club,
  player: Player,
  fee: number = 0,
): boolean {
  if (club.team.players.length >= MAX_SQUAD_SIZE) return false;
  const reserve = fee > 0 ? salaryReserve(club) : 0;
  if (fee > 0 && club.balance - fee < reserve) return false;

  if (fee > 0) {
    club.balance -= fee;
  }

  addPlayerToClub(club, player);
  return true;
}

/**
 * Checks if a player can be removed from a club without violating rules.
 */
export function canRemovePlayer(club: Club, player: Player): boolean {
  if (club.team.players.length <= MIN_SQUAD_SIZE) return false;

  if (player.primaryPosition === Position.GK) {
    const gkCount = club.team.players.filter(
      (p) => p.primaryPosition === Position.GK,
    ).length;
    if (gkCount <= 1) return false;
  }

  return true;
}

function removePlayerFromClub(club: Club, player: Player): void {
  const newPlayers = club.team.players.filter((p) => p.id !== player.id);
  const newLineup = club.team.startingLineup.filter((id) => id !== player.id);
  const mutableLineup = [...newLineup];

  // If we removed a starter, fill from reserves
  if (
    club.team.startingLineup.includes(player.id) &&
    newPlayers.length >= 11
  ) {
    const lineupSet = new Set(mutableLineup);
    const reserve = newPlayers.find((p) => !lineupSet.has(p.id));
    if (reserve != null) {
      mutableLineup.push(reserve.id);
    }
  }

  club.team = {
    ...club.team,
    players: newPlayers,
    startingLineup: mutableLineup,
  };
}

function addPlayerToClub(club: Club, player: Player): void {
  const newPlayers = [...club.team.players, player];
  const newLineup = [...club.team.startingLineup];

  if (newLineup.length < 11) {
    newLineup.push(player.id);
  }

  club.team = {
    ...club.team,
    players: newPlayers,
    startingLineup: newLineup,
  };
}
